File: mouve/ViewPrototype/NodeController.py
from PySide.QtCore import *
from PySide.QtGui import *
from NodeScene import *

controller = None
DEBUG_LINKS = False

class NodeController(QObject):

	# These ones goes to model
	_registeredNodeTypes = ['Image from disk', 'Gaussian blur', \
				'Preview window', 'Grayscale', 'Mixture of Gaussians']
	_nodeId = 0

	def __init__(self, parent=None):
		QObject.__init__(self, parent)

		self._nodeViews = {}
		self._linkViews = []

		self._scene = NodeScene(self)
		# !TODO: Temporary
		self._scene.setSceneRect(-200,-200,1000,600)

		# !TODO: Move to some initialize method
		self._view = NodeEditorView()
		self._view.setScene(self._scene)
		self._view.setGeometry(100, 50, 1200, 800)
		self._view.contextMenu.connect(self.contextMenu)
		self._view.show()

		self._addNodesActions = []
		for item in enumerate(self._registeredNodeTypes):
			a = QAction(item[1], self)
			a.setData(item[0])
			self._addNodesActions.append(a)

	'''
	Zwraca widok wezla o podanym kluczu ID
	'''
	def nodeView(self, key):
		return self._nodeViews[key]

	'''
	Dodaje nowy widok wezla
	'''
	def addNodeView(self, key, title):
		node = NodeView(title)
		node.setData(NodeDataIndex.NodeKey, key)
		#node.installEventFilter(self.nodeManager)
		self._nodeViews[key] = node
		self._scene.addItem(node)
		return node

	''' 
	Usuwa widok wezla o podanym kluczu ID
	'''
	def deleteNodeView(self, key):
		print 'deleting node with key:', key
		# if key in self._nodeViews:
		# 	print 'deleting node with key: {0}'.format(key)
		# 	nodeView = self._nodeViews.pop(key)
		# 	# !TODO: Nie pamietam juz czy logika sie tym zajmie czy nie
		# 	#         - chodzi o usuniecie wiszacych polaczen
		# 	lvs = filter(lambda lv: lv.connecting(nodeView), self._linkViews)
		# 	for link in lvs:
		# 		link.dtor()
		# 		self._linkViews.remove(link)

		# 	self._scene.removeItem(nodeView)
		# 	return True
		# else:
		# 	print 'node with key: {0} not found'.format(key)
		# 	return False

	'''
	Zwraca liste polaczen wchodzacych do danego wezla
	'''
	def inputLinkViews(self, nodeKey):
		if nodeKey in self._nodeViews:
			nv = self._nodeViews[nodeKey]
			
			return filter(lambda lv: lv.inputConnecting(nv), self._linkViews)
		return None

	'''
	Zwraca liste polaczen wychodzacych z danego wezla
	'''
	def outputLinkViews(self, nodeKey):
		if nodeKey in self._nodeViews:
			nv = self._nodeViews[nodeKey]
			
			return filter(lambda lv: lv.outputConnecting(nv), self._linkViews)
		return None

	'''
	Tworzy nowe (wizualne) polaczenie miedzy danymi gniazdami
	'''
	def linkNodeViews(self, fromSocketView, toSocketView):
		# TODO!: To bedzie w logice
		# Check if we aren't trying to create already existing connection
		for linkView in self._linkViews:
			if linkView.connects(fromSocketView, toSocketView):
				QMessageBox.critical(None, 'NodeView', 'Connection already exists')
				return

		link = NodeLinkView(fromSocketView, toSocketView, None)
		link.setDrawDebug(DEBUG_LINKS)
		fromSocketView.addLink(link)
		toSocketView.addLink(link)

		self._linkViews.append(link)
		#link.installEventFilter(self.nodeManager)
		self._scene.addItem(link)

	'''
	Usuwa polaczenie (wizualne) miedzy danymi gniazdami
	'''
	def unlinkNodeViews2(self, fromSocketView, toSocketView):
		for linkView in self._linkViews:
			if linkView.connects(fromSocketView, toSocketView):
				self._scene.removeItem(linkView)
				self._linkViews.remove(linkView)
				return True
		return False

	def unlinkNodeViews(self, linkView):
		pass

	# Normalnie, za to odpowiedzialna jest logika
	def _addNode(self, nodeTypeId, scenePos):
		if nodeTypeId < len(self._registeredNodeTypes):
			title = self._registeredNodeTypes[nodeTypeId]
			nodeId = self._nodeId
			node = self.addNodeView(nodeId, title)
			node.setPos(scenePos)
			if title == 'Image from disk':
				node.addSocketView(0, 'Output', True)
			elif title == 'Gaussian blur':
				node.addSocketView(0, 'Input')
				node.addSocketView(1, 'Sigma')
				node.addSocketView(0, 'Output', True)
			elif title == 'Preview window':
				node.addSocketView(0, 'Input')
			elif title == 'Grayscale':
				node.addSocketView(0, 'Input')
				node.addSocketView(0, 'Output', True)
			elif title == 'Mixture of Gaussians':
				node.addSocketView(0, 'Frame')
				node.addSocketView(0, 'Foreground mask', True)
			self._nodeId += 1

	'''
	Reakcja na opuszczenie nowego polaczenia
	'''
	@Slot(NodeSocketView, NodeSocketView)
	def draggingLinkDropped(self, fromSocketView, toSocketView):
		self.linkNodeViews(fromSocketView, toSocketView)

	@Slot(NodeSocketView)
	def draggingLinkStarted(self, fromSocketView):
		self._scene.dragging = True
		#print 'dragging started from', fromSocketView

	@Slot(NodeSocketView)
	def draggingLinkStopped(self, fromSocketView):
		self._scene.dragging = False
		#print 'dragging stopped'

	'''
	Reakcja na probe pokazania menu kontekstowego pochodzacego z widoku sceny
	'''
	@Slot(QPoint, QPointF)
	def contextMenu(self, globalPos, scenePos):
		items = self._scene.items(scenePos, Qt.ContainsItemShape, Qt.AscendingOrder)
		# If we clicked onto empty space
		if not items:
			menu = QMenu()
			menuAddNode = menu.addMenu('Add node')
			for a in self._addNodesActions:
		 		menuAddNode.addAction(a)
		 	ret = menu.exec_(globalPos)
		 	if ret is not None:
		 		self._addNode(ret.data(), scenePos)
		else:
			for item in items:
				if item.type() == NodeView.Type:
					menu = QMenu()
					action = QAction('Delete node', None)
					action.setData(10)
					menu.addAction(action)
					ret = menu.exec_(globalPos)
					if ret is not None and isinstance(ret, QAction):
						#print item.nodeKey()
						self.deleteNodeView(item.nodeKey())
					break

	def _init_sample_scene(self):
		self._addNode(self._registeredNodeTypes.index('Image from disk'), QPoint(-250,0))
		self._addNode(self._registeredNodeTypes.index('Grayscale'), QPoint(-40,50))
		self._addNode(self._registeredNodeTypes.index('Gaussian blur'), QPoint(150,100))
		self._addNode(self._registeredNodeTypes.index('Mixture of Gaussians'), QPoint(350,150))
		self._addNode(self._registeredNodeTypes.index('Preview window'), QPoint(600,100))

		#self.linkNodeViews(sv1,sv2)
		#self.linkNodeViews(sv1,sv5)
		#self.linkNodeViews(sv3,s4v)

		#self.deleteNodeView(1)
		#self.unlinkNodeViews2(sv1, sv2)

File: mouve/NodeController.cpp
#include "NodeController.h"
#include "NodeScene.h"
#include "NodeEditorView.h"
#include "NodeSocketView.h"

#include <QMenu>
#include <QAction> 
#include <QDebug>
#include <QMessageBox>

static QString registeredNodeTypes[] = {
	"Image from disk",
	"Gaussian blur",
	"Preview window",
	"Grayscale",
	"Mixture of Gaussians"
};
static int nodeId = 0;

template<> NodeController* Singleton<NodeController>::_singleton = nullptr;

NodeController::NodeController(QObject* parent)
	: QObject(parent)
	, mScene(new NodeScene(this))
	, mView(new NodeEditorView)
{
	// !TODO: Temporary
	mScene->setSceneRect(-200,-200,1000,600);
	// Qt bug concering scene->removeItem ?? Seems to fixed it
	mScene->setItemIndexMethod(QGraphicsScene::NoIndex);

	// !TODO: Move to some initialize method
	mView->setScene(mScene);
	mView->setGeometry(100, 50, 1200, 800);
	connect(mView, SIGNAL(contextMenu(QPoint,QPointF)),
		this, SLOT(contextMenu(QPoint,QPointF)));
	connect(mView, SIGNAL(keyPress(QKeyEvent*)),
		this, SLOT(keyPress(QKeyEvent*)));
	mView->show();

	for(int i = 0; i < sizeof(registeredNodeTypes)/sizeof(QString); ++i)
	{
		QAction* a = new QAction(registeredNodeTypes[i], this);
		a->setData(i);
		mAddNodesActions.append(a);
	}
}

NodeController::~NodeController()
{
	foreach(NodeLinkView* link, mLinkViews)
		delete link;
	mLinkViews.clear();

	foreach(NodeView* node, mNodeViews.values())
		delete node;
	mNodeViews.clear();

	delete mView;
}

NodeView* NodeController::nodeView(quint32 nodeKey)
{
	return mNodeViews[nodeKey];
}

NodeView* NodeController::addNodeView(quint32 nodeKey, const QString& title)
{
	NodeView* node = new NodeView(title);
	node->setData(NodeDataIndex::NodeKey, nodeKey);
	mNodeViews[nodeKey] = node;
	mScene->addItem(node);
	return node;
}

bool NodeController::deleteNodeView(quint32 nodeKey)
{
	qDebug() << "deleting node view with nodeKey:" << nodeKey;

	auto it = mNodeViews.find(nodeKey);
	if(it != mNodeViews.end())
	{
		NodeView* node = *it;

		// Remove incoming and outgoing links
		QMutableListIterator<NodeLinkView*> itl(mLinkViews);
		while(itl.hasNext())
		{
			NodeLinkView* link = itl.next();

			if(link->inputConnecting(node) ||
			   link->outputConnecting(node))
			{
				mScene->removeItem(link);
				itl.remove();
				delete link;
			}
		}

		// Remove from the scene
		mScene->removeItem(node);
		// Remove from container
		mNodeViews.erase(it);
		// Remove at last node view itself
		node->deleteLater();
		return true;
	}
	return false;
}

bool NodeController::deleteNodeView(NodeView* nodeView)
{
	return deleteNodeView(
		quint32(nodeView->data(NodeDataIndex::NodeKey).toInt()));
}

void NodeController::linkNodeViews(NodeSocketView* from, NodeSocketView* to)
{
	// TODO!: To bedzie w logice
	// Check if we aren't trying to create already existing connection
	for(auto it = mLinkViews.begin(); it != mLinkViews.end(); ++it)
	{
		if((*it)->connects(from, to))
		{
			QMessageBox::critical
				(nullptr, "NodeView", "Connection already exists");
			return;
		}
	}

	NodeLinkView* link = new NodeLinkView(from, to, nullptr);
	link->setDrawDebug(DEBUG_LINKS);
	from->addLink(link);
	to->addLink(link);
	mLinkViews.push_back(link);
	mScene->addItem(link);
}

bool NodeController::unlinkNodeViews(NodeSocketView* from, NodeSocketView* to)
{
	for(int i = 0; i < mLinkViews.size(); ++i)
	{
		if(mLinkViews[i]->connects(from, to))
		{
			NodeLinkView* link = mLinkViews[i];
			mScene->removeItem(link);
			mLinkViews.removeAt(i);
			delete link;			
			return true;
		}
	}

	return false;
}

bool NodeController::unlinkNodeViews(NodeLinkView* link)
{
	if(mLinkViews.removeOne(link))
	{
		mScene->removeItem(link);
		delete link;
		return true;
	}
	return false;
}

void NodeController::addNode(quint32 nodeTypeId, const QPointF& scenePos)
{
	if(nodeTypeId < sizeof(registeredNodeTypes)/sizeof(QString))
	{
		QString title = registeredNodeTypes[nodeTypeId];

		NodeView* node = addNodeView(nodeId, title);
		node->setPos(scenePos);

		if(title == "Image from disk")
		{
			node->addSocketView(0, "Output", true);
		}
		else if(title == "Gaussian blur")
		{
			node->addSocketView(0, "Input");
			node->addSocketView(1, "Sigma");
			node->addSocketView(0, "Output", true);
		}
		else if(title == "Preview window")
		{
			node->addSocketView(0, "Input");
		}
		else if(title == "Grayscale")
		{
			node->addSocketView(0, "Input");
			node->addSocketView(0, "Output", true);
		}
		else if(title == "Mixture of Gaussians")
		{
			node->addSocketView(0, "Frame");
			node->addSocketView(0, "Foreground mask", true);
		}

		nodeId += 1;
	}
}

void NodeController::initSampleScene()
{
	addNode(0, QPointF(-250,0));
	addNode(3, QPointF(-40,50));
	addNode(1, QPointF(150,100));
	addNode(4, QPointF(350,150));
	addNode(2, QPointF(600,100));
}

void NodeController::draggingLinkDropped(QGraphicsWidget* from, 
	QGraphicsWidget* to)
{
	linkNodeViews(static_cast<NodeSocketView*>(from), static_cast<NodeSocketView*>(to));
}

void NodeController::draggingLinkStarted(QGraphicsWidget* from)
{
	mScene->setDragging(true);
}

void NodeController::draggingLinkStopped(QGraphicsWidget* from)
{
	mScene->setDragging(false);
}

void NodeController::contextMenu(const QPoint& globalPos,
	const QPointF& scenePos)
{
	QList<QGraphicsItem*> items = mScene->items(scenePos, 
		Qt::ContainsItemShape, Qt::AscendingOrder);
	//If we clicked onto empty space
	if(items.isEmpty())
	{
		QMenu menu;
		QMenu* menuAddNode = menu.addMenu("Add node");
		foreach(QAction* a, mAddNodesActions)
			menuAddNode->addAction(a);
		QAction* ret = menu.exec(globalPos);
		if(ret != nullptr)
			addNode(ret->data().toInt(), scenePos);
	}
	else
	{
		for(int i = 0; i < items.size(); ++i)
		{
			if(items[i]->type() == NodeView::Type)
			{
				QMenu menu;
				QAction action("Delete node", nullptr);
				action.setData(10);
				menu.addAction(&action);
				QAction* ret = menu.exec(globalPos);
				if(ret != nullptr)
				{
					NodeView* node = static_cast<NodeView*>(items[i]);
					deleteNodeView(node->nodeKey());
				}
				return;
			}
		}
	}
}

void NodeController::keyPress(QKeyEvent* event)
{
	if(event->key() == Qt::Key_Delete)
	{
		QList<QGraphicsItem*> itemsPing = mScene->selectedItems();
		QList<QGraphicsItem*> itemsPong;
		mScene->clearSelection();

		for(int i = 0; i < itemsPing.size(); ++i)
		{
			if(itemsPing[i]->type() == NodeLinkView::Type)
			{
				NodeLinkView* link = static_cast<NodeLinkView*>(itemsPing[i]);
				unlinkNodeViews(link);
			}
			else
			{
				itemsPong.append(itemsPing[i]);
			}
		}
		itemsPing.clear();

		for(int i = 0; i < itemsPong.size(); ++i)
		{
			if(itemsPong[i]->type() == NodeView::Type)
			{
				NodeView* node = static_cast<NodeView*>(itemsPong[i]);
				deleteNodeView(node);
			}
		}

		event->accept();
	}
}

